services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryManager.java

```diff
@@ -0,0 +1,166 @@
+                             DefaultCompletedUploadDirectory.builder()
+                                                            .failedUploads(failedUploads)
+                                                            .successfulUploads(uploads)
+                                                            .build()));
+        return new DefaultUploadDirectory(returnFuture);
+    }
+
+    private CompletableFuture<CompletedUpload> uploadSingleFile(UploadDirectoryRequest uploadDirectoryRequest,
+                                                                List<CompletedUpload> uploads,
+                                                                List<FailedUpload> failedUploads,
+                                                                Path path) {
+        int nameCount = uploadDirectoryRequest.sourceDirectory().getNameCount();
+        UploadRequest uploadRequest = constructUploadRequest(uploadDirectoryRequest, nameCount, path);
+        CompletableFuture<CompletedUpload> future = uploadFunction.apply(uploadRequest)
+                                                                  .completionFuture();
+        future.whenComplete((r, t) -> {
+            if (t != null) {
+                failedUploads.add(DefaultFailedUpload.builder()
+                                                     .exception(t)
+                                                     .path(path)
+                                                     .build());
+            } else {
+                uploads.add(r);
+            }
+        });
+        return future;
+    }
+
+    private Stream<Path> listFiles(Path directory, UploadDirectoryRequest request) {
+
+        try {
+            boolean recursive = transferConfiguration.resolveUploadDirectoryRecursive(request);
+
+            if (!recursive) {
+                return Files.list(directory).filter(Files::isRegularFile);
+            }
+
+            boolean followSymbolicLinks = transferConfiguration.resolveUploadDirectoryFollowSymbolicLinks(request);
+
+            int maxDepth = transferConfiguration.resolveUploadDirectoryMaxDepth(request);
+
+            if (followSymbolicLinks) {
+                return Files.walk(directory, maxDepth, FileVisitOption.FOLLOW_LINKS)
+                            .peek(p -> log.trace(() -> "Processing path: " + p))
+                            .filter(Files::isRegularFile);
+            }
+
+            return Files.walk(directory, maxDepth)
+                        .peek(p -> log.trace(() -> "Processing path: " + p))
+                        .filter(path -> Files.isRegularFile(path, LinkOption.NOFOLLOW_LINKS));
+
+        } catch (IOException e) {
+            throw SdkClientException.create("Failed to list files under the provided directory", e);
+        }
+    }
+
+    private static String processPrefix(String prefix) {
+        if (StringUtils.isEmpty(prefix)) {
+            return "";
+        }
+        return prefix.endsWith("/") ? prefix : prefix + "/";
```

Comment:
S3 delimiters are configurable: https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html

Should we support that here, instead of hard-coding the delimiter as /?